codec/src/main/java/io/netty/handler/codec/compression/ZstdEncoder.java

```diff
@@ -0,0 +1,326 @@
+import static io.netty.handler.codec.compression.ZstdConstants.BLOCK_TYPE_COMPRESSED;
+import static io.netty.handler.codec.compression.ZstdConstants.ZSTD_NUMBER;
+import static io.netty.handler.codec.compression.ZstdConstants.TOKEN_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.COMPRESSED_LENGTH_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.DECOMPRESSED_LENGTH_OFFSET;
+import static io.netty.handler.codec.compression.ZstdConstants.DEFAULT_CHECKSUM_OFFSET;
+import static io.netty.util.internal.ThrowableUtil.unknownStackTrace;
+
+/**
+ *  Compresses a {@link ByteBuf} using the Zstandard algorithm.
+ *  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *     * * * * * * * * * * * * *
+ *  * ZSTD number * CompressType *  Compressed *  Decompressed *  Checksum *  +  *  ZSTD compressed *
+ *  *       *              *    length   *     length    *           *     *      block      *
+ *  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *     * * * * * * * * * * * * *
+ */
+public class ZstdEncoder extends MessageToByteEncoder<ByteBuf> {
+
+    private static final EncoderException ENCODE_FINSHED_EXCEPTION = unknownStackTrace(new EncoderException(
+                    new IllegalStateException("encode finished and not enough space to write remaining data")),
+            ZstdEncoder.class, "encode");
+    private final int blockSize;
+    private final Adler32 checksum = new Adler32();
+    private final int compressionLevel;
+    private ByteBuf buffer;
+    private final int maxEncodeSize;
+    private volatile boolean finished;
+    private volatile ChannelHandlerContext ctx;
+
+    public ZstdEncoder() {
+        this(DEFAULT_BLOCK_SIZE, MAX_BLOCK_SIZE);
+    }
+
+    public ZstdEncoder(int blockSize, int maxEncodeSize) {
+        super(true);
+        compressionLevel = compressionLevel(blockSize);
+        this.blockSize = blockSize;
+        this.maxEncodeSize = ObjectUtil.checkPositive(maxEncodeSize, "maxEncodeSize");
+        finished = false;
+    }
+
+    protected static int compressionLevel(int blockSize) {
+        if (blockSize < MIN_BLOCK_SIZE || blockSize > MAX_BLOCK_SIZE) {
+            throw new IllegalArgumentException(String.format(
+                    "blockSize: %d (expected: %d-%d)", blockSize, MIN_BLOCK_SIZE, MAX_BLOCK_SIZE));
+        }
+        int compressionLevel = 32 - Integer.numberOfLeadingZeros(blockSize - 1); // ceil of log2
+        compressionLevel = Math.max(0, compressionLevel - COMPRESSION_LEVEL_BASE);
+        return compressionLevel;
+    }
+
+    @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return allocateBuffer(ctx, msg, preferDirect, true);
+    }
+
+    private ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect,
+                                   boolean allowEmptyReturn) {
```

Comment:
allowEmptyReturn is unused, because it needs to be distinguished from the `io.netty.handler.codec.MessageToByteEncoder.allocateBuffer(ChannelHandlerContext ctx, I msg, boolean preferDirect) `method